File: java/src/main/java/datastructures/LinkedList.java
package datastructures;

public class LinkedList<T> {

    private int size;
    private Node<T> head;

    /**
     * O(1)
     *
     * without loops
     *
     * @param data
     */
    public void addFront(final T data) {
        //create a new new Node
        final var newNode = new Node<T>(data);

        if (head != null) {
            //Set new new to head then pointer to older head
            newNode.next = this.head;

            //Current head, point to new node
        }
        this.head = newNode;
        increaseSize();
    }

    /**
     * O(n)
     *
     * @param data
     */
    public void addBack(final T data) {
        final var newTail = new Node<T>(data);

        if (head == null){
            head = newTail;
        }else {
            //Set new TAIL
            toTail().next = newTail; // --> O(n)
        }
        increaseSize();
    }

    /**
     * O(1)
     *
     * @return
     */
    public T getFirst() {
        if(this.head == null){
            throw new IllegalStateException("Head is NULL");
        }
        return this.head.data;
    }

    /**
     * O(n)
     *
     * Get last is same as Get Tail
     * @return Tail
     */
    public T getLast() {
        if(this.head == null){
            throw new IllegalStateException("Head is NULL");
        }

        return toTail().data; // --> O(n)
    }

    public int getSize() {
        return size;
    }

    /**
     * O(n)
     *
     * Size by brute force
     * @return
     */
    public int count() {
        if (head == null) {
            return 0;
        }else {
            int count = 1;
            var current = head;

            //while we are not at TAIL
            while (current.next != null) { // --> O(n)
                current = current.next;
                count++;
            }
            return count;
        }
    }

    public void clear() {
        this.size = 0;
        this.head = Node.EMPTY; //let the garbage collector work. Clean all nodes
    }

    /**
     * O(n)
     *
     * @param value
     */
    public void delete(final T value) {
        if (head == null) {
            return;
        }
        var current = head;
        var removed = false;

        if (value.equals(head.data)){
            head = head.next;
            removed = true;
        } else {

            //while we are not at TAIL
            while (current.next != null) { // --> 0(n)
                if (current.next.data.equals(value)) {
                    current.next = current.next.next;
                    removed = true;
                    break;
                }
                current = current.next;
            }
        }
        decreaseSizeIf(removed);
    }

    /**
     * O(n)
     *
     * Start at the Head
     * Walking down to the end
     * @return tail
     */
    private Node<T> toTail() {
        var current = head;

        //while we are not at TAIL
        while (current.next != null) {
            current = current.next;
        }

        return current;
    }

    private void increaseSize() {
        this.size++;
    }

    private void decreaseSizeIf(final boolean removed) {
        if (removed) {
            this.size--;
        }
    }

    private static class Node<T> {
        private static final Node EMPTY = null;
        private T data;
        private Node<T> next;

        public Node(final T data) {
            this.data = data;
        }
    }
}

File: go/datastructures/linked_list.go
package datastructures

type LinkedList struct {
	head *Node
	size int
}

func (list *LinkedList) New() LinkedList {
	return LinkedList{
		head: nil,
		size: 0,
	}
}

func (list *LinkedList) AddFront(data interface{}) {
	/*newNode := &node{
		data: data,
		next: list.head,
	}

	list.head = newNode*/

	list.head = &Node{
		Data: data,
		Next: list.head,
	}
	list.size++
}

func (list *LinkedList) AddBack(data interface{}) {

	newTail := &Node{
		Data: data,
		Next: nil,
	}
	currentTail := list.tail()

	if currentTail != nil {
		currentTail.Next = newTail
	} else {
		list.head = newTail
	}
	list.size++
}

func (list *LinkedList) First() interface{} {
	return list.head.Data
}

func (list *LinkedList) Last() interface{} {
	return list.tail().Data
}

func (list *LinkedList) tail() *Node {
	node := list.head

	for node != nil && node.Next != nil {
		node = node.Next
	}
	return node
}

func (list *LinkedList) Delete(value interface{}) {
	current := list.head

	//if head
	if value == current.Data {
		list.head = current.Next
		list.size--
		return
	}
	for current != nil && current.Next != nil {
		if value == current.Next.Data {
			current.Next = current.Next.Next
		}
		current = current.Next
	}
	list.size--
}


